src/inmem/immutable.rs

```diff
@@ -127,25 +143,44 @@ where
     }
 }
 
-pub(crate) struct ImmutableScan<'iter, R>
+pub(crate) enum ImmutableScan<'iter, R>
 where
     R: Record,
 {
-    range: Range<'iter, Ts<<R::Schema as Schema>::Key>, u32>,
-    record_batch: &'iter RecordBatch,
-    projection_mask: ProjectionMask,
+    Forward {
+        range: Range<'iter, Ts<<R::Schema as Schema>::Key>, u32>,
+        record_batch: &'iter RecordBatch,
+        projection_mask: ProjectionMask,
+    },
+    Reverse {
+        range: Rev<Range<'iter, Ts<<R::Schema as Schema>::Key>, u32>>,
+        record_batch: &'iter RecordBatch,
+        projection_mask: ProjectionMask,
+    },
 }
 
 impl<'iter, R> ImmutableScan<'iter, R>
 where
     R: Record,
 {
-    fn new(
+    fn new_forward(
```

Comment:
Maybe we can rename this to something like: 
```suggestion
    fn forward_scan(
```